File: 7 欧治/ouyee.js
// 环境
// content文件
// content = "6PWkevrej8oSnK4bjyH2EMA_i5umaKsc0Sa6NV1SfyIE1UwCEMYpM1aBNeKwKcTsh6QRiTukdOf4rLoCVWcNUqJTc.xvRzbvby830fRjYfEIhLN6Nyg.miFNFKFKiDcKDxSuJjYLz7QfUYR2_1gnD.v1blOGu6g1pqUSmeGOBhVwbFnv1KeC5Id_4zqdwhBcQGrPgmIvkyMEaedN3MoeTtLJMK8x_DpsUFePRhKOsSa"
content = 'metaContent'
// 放置环境与proxy检测的地儿

window = self = top = global;

// window.parseFloat = {}
// window.webkitRequestFileSystem = {} // 不要瞎补这些多出来看似有用实则完全无用纯干扰人的东西
window.addEventListener = function (){}
window.setInterval = function () {}
window.setTimeout = function () {}
window.ActiveXObject = undefined

location = {
    "ancestorOrigins": {},
    "href": "https://www.ouyeel.com/steel/search?keySearch=&pageIndex=0&pageSize=50",
    "origin": "https://www.ouyeel.com",
    "protocol": "https:",
    "host": "www.ouyeel.com",
    "hostname": "www.ouyeel.com",
    "port": "",
    "pathname": "/steel/search",
    "search": "?keySearch=&pageIndex=0&pageSize=50",
    "hash": ""
}

navigator = {
    userAgent:'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36 '
}

i = {}

div = {
    getElementsByTagName(ele) {
        // console.log('div getElementsByTagName',ele)
        if (ele === 'i') {
            return []
        } else{
            return []
        }
    }
}

head = {
    removeChild() {

    }
}

script = {
    getAttribute(ele) {
        // console.log('script getAttribute',ele)
        if (ele === 'r') {
            return 'm'
        }
    },
    parentElement: head,
}
meta = {
        getAttribute(ele) {
        // console.log('meta getAttribute',ele)
        if (ele === 'r') {
            return 'm'
        }
    },
    parentNode:head,
    content:content
}

form = {}

document = {
    createElement(ele) {
        console.log('document createElement',ele)
        if (ele === 'div') {
            return div
        } else if (ele === "form") {
            return form
        } else {
            return {} // 当有其他情况时，让其返回空数组而非undefined
        }
    },
    appendChild(ele) {
        // console.log('document appendChild',ele)
        return {}
    },
    removeChild(ele) {
        // console.log('document removeChild',ele)
        return {}
    },
    getElementsByTagName(ele) {
        // console.log('document getElementsByTagName',ele)
        if (ele === 'script') {
            return [script, script]
        }else if(ele==='meta'){
            return [meta,meta]
        }else if(ele==='base'){
            return []
        }else{
            return []
        }
    },
    getElementById() {
        // console.log('document getElementById',ele)
        // 不重要的值
    }
}



// 源码

"ts_js"
;
"auto_js"
;


// api
function get_cookie(){
    return document.cookie
}

// console.log(get_cookie())

File: 7 欧治/ouyee.py
import json
import random
import sys
import time

from lxml import etree
import requests

# 为subprocess在使用execjs的情况下统一设置utf8，同时不让其因为改变影响别的用到subprocess的地方
from functools import partial
import subprocess
my_popen = partial(subprocess.Popen, encoding='utf-8')
import execjs
from loguru import logger
execjs._external_runtime.Popen = my_popen


# 1.初始化各种参数
session = requests.session()
header = {
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36'
}
url = 'https://www.ouyeel.com/search-ng/commoditySearch/queryCommodityResult'


# 2.第一次发起请求访问瑞数页，获取三大件
def surf_1():
    res = session.get(url, headers=header).text
    tree = etree.HTML(res)
    metaContent = tree.xpath('//meta/@content')[-1]
    ts_script = tree.xpath('//script/text()')[0]
    auto_url = 'https://www.ouyeel.com' + tree.xpath('//script/@src')[0]
    auto_script = session.get(auto_url).text
    return metaContent, ts_script, auto_script


# 3.将三大件动态替换并使用其生成cookie2
def get_cookie2(metaContent, ts_script, auto_script):
    with open('ouyee.js', encoding='utf8') as f:
        js_code = f.read()
    # 原文就是双引号，所以要把那个双引号带着包裹一块替换
    js_code = js_code.replace('metaContent', metaContent).replace('"ts_js"', ts_script).replace('"auto_js"',
                                                                                                auto_script)
    js_compile = execjs.compile(js_code)
    js_cookie_key_value = js_compile.call('get_cookie').split(';')[0].split('=')
    # print(js_cookie_key_value)
    session.cookies.update({js_cookie_key_value[0]: js_cookie_key_value[1]})
    # print(session.cookies.get_dict())


# 4.更新c2并第二次发起请求访问数据页,并做数据清洗
def surf_2(page_index):
    data = {
        "criteriaJson": '{"pageIndex":%d,"pageSize":50,"industryComponent":null,"channel":null,"productType":null,'
                        '"sort":null,"warehouseCode":null,"key_search":null,"is_central":null,"searchField":null,'
                        '"companyCode":null,"inquiryCategory":null,"inquirySpec":null,"provider":null,'
                        '"packCodes":null,"shopCode":null,"steelFactory":null,"resourceIds":null,"jsonParam":{},'
                        '"excludeShowSoldOut":null}' % page_index
    }
    response = session.post(url, headers=header, data=data).json()
    logger.info(f'正在打印第{page_index}页产品列表...')
    ret = response['resultList']
    ret = json.loads(ret)
    for i in ret:
        print(i)


# 5.编写爬虫整体逻辑
def main():
    for page in range(1, 8):
        metaContent, ts_script, auto_script = surf_1()
        get_cookie2(metaContent, ts_script, auto_script)
        surf_2(page)
        session.cookies.clear()  # 瑞数特点，需要不断清空cookie并给其新动态cookie值
        time.sleep(random.uniform(0.5, 0.8))


# 5.执行
if __name__ == '__main__':
    main()
